framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,563 @@
+        }
+    }
+
+    /**
+     * Prints all the fields of a given class
+     *
+     * @param aClass the class whose fields should be printed
+     * @param classname the simple name of the class, used to print the type of "this"
+     * @param printWriter the writer on which to print the fields
+     */
+    private static void printFields(
+            AClassWrapper aClass, String classname, PrintWriter printWriter) {
+        for (Map.Entry<String, AFieldWrapper> fieldEntry : aClass.getFields().entrySet()) {
+            String fieldName = fieldEntry.getKey();
+            AFieldWrapper aField = fieldEntry.getValue();
+            printWriter.print(INDENT);
+            printWriter.print(formatAField(aField, fieldName, classname));
+            printWriter.println(";");
+            printWriter.println();
+        }
+    }
+
+    /**
+     * Prints a method signature in stub file format (i.e., without a method body).
+     *
+     * @param aMethodWrapper the method to print
+     * @param basename the simple name of the containing class. Used only to determine if the method
+     *     being printed is the constructor of an inner class.
+     * @param printWriter where to print the method signature
+     */
+    private static void printMethodSignature(
+            AMethodWrapper aMethodWrapper, String basename, PrintWriter printWriter) {
+
+        AMethod aMethod = aMethodWrapper.getAMethod();
+
+        printWriter.print(INDENT);
+        printWriter.print(formatAnnotations(aMethod.returnType.tlAnnotationsHere));
+        // Needed because AMethod stores the name with the parameters, to differentiate
+        // between different methods in the same class with the same name.
+        String methodName = aMethod.methodName.substring(0, aMethod.methodName.indexOf("("));
+        // Use Java syntax for constructors.
+        if ("<init>".equals(methodName)) {
+            // Constructor names cannot contain dots, if this is an inner class.
+            methodName =
+                    basename.contains(".")
+                            ? basename.substring(basename.lastIndexOf('.') + 1)
+                            : basename;
+        } else {
+            // This isn't a constructor, so add a return type.
+            // Note that the stub file format doesn't require this to be correct,
+            // so it would be acceptable to print "java.lang.Object" for every
+            // method. A better type is printed if one is available to improve
+            // the readability of the resulting stub file.
+            String returnType = aMethodWrapper.getReturnType();
+            printWriter.print(returnType);
+            printWriter.print(" ");
+        }
+        printWriter.print(methodName);
+        printWriter.print("(");
+
+        StringJoiner parameters = new StringJoiner(", ");
+
+        if (!aMethod.receiver.type.tlAnnotationsHere.isEmpty()
+                || !aMethod.receiver.type.innerTypes.isEmpty()) {
```

Comment:
This isn't testing for annotations, just whether the types exist.  Should it look for annotations?